lib/streamlit/hashing.py

```diff
@@ -78,22 +97,59 @@ class HashStacks(object):
     """
 
     def __init__(self):
-        self.stacks = collections.defaultdict(list)  # type: Dict[int, List[int]]
+        self._stack = (
+            collections.OrderedDict()
+        )  # type: collections.OrderedDict[int, List[int]]
+
+        # The reason why we're doing this hashing, for debug purposes.
+        self.hash_reason = None
+
+        # Either a function or a code block, depending on whether the reason is
+        # due to hashing part of a function (i.e. body, args, output) or an
+        # st.Cache codeblock.
+        self.hash_source = None
 
     def push(self, val):
-        thread_id = threading.current_thread().ident or -1
-        self.stacks[thread_id].append(id(val))
+        self._stack[id(val)] = val
 
     def pop(self):
-        thread_id = threading.current_thread().ident or -1
-        self.stacks[thread_id].pop()
+        return self._stack.popitem()
 
     def __contains__(self, val):
-        thread_id = threading.current_thread().ident or -1
-        return id(val) in self.stacks[thread_id]
+        return id(val) in self._stack
 
+    def pretty_print(self):
```

Comment:
Could an acceptable optimization be to remove all the types that we have an internal hash_func for?

Reply:
Hmmm... I have to think about that.

You're right that those aren't useful to users, but I think they might be useful when we're debugging an issue sent by a user.

I vote to leave those in for now, and figure out whether to remove later based on our debugging experience. Left a note in the code.